jdbc/src/test/java/org/apache/iotdb/jdbc/IoTDBQueryResultSetTest.java

```diff
@@ -135,7 +123,11 @@ public void before() throws Exception {
     when(connection.isClosed()).thenReturn(false);
```

Comment:
As I understand it, this test can not cover the bug you want to solve. 
You can add a test contains one client with two same query statements.

Besides, do you fix the bug that a statement.close could close the whole session?

Reply:
> As I understand it, this test can not cover the bug you want to solve.

+1. An integration test under the iotdb module to cover the bug is the best. I am interested in this bug so I write the following example test in the master branch and really catch the bugs in both the query results and the close of resultSets.

```
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.iotdb.db.integration;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.sql.Statement;
import org.apache.iotdb.db.conf.IoTDBDescriptor;
import org.apache.iotdb.db.service.IoTDB;
import org.apache.iotdb.db.utils.EnvironmentUtils;
import org.apache.iotdb.jdbc.Config;
import org.apache.iotdb.tsfile.common.conf.TSFileConfig;
import org.apache.iotdb.tsfile.common.conf.TSFileDescriptor;
import org.junit.AfterClass;
import org.junit.Assert;
import org.junit.BeforeClass;
import org.junit.Test;

/**
 * Notice that, all test begins with "IoTDB" is integration test. All test which will start the IoTDB server should be
 * defined as integration test.
 */
public class IoTDBMultiStatementsIT {

  private static IoTDB daemon;

  private static boolean testFlag = Constant.testFlag;
  private static TSFileConfig tsFileConfig = TSFileDescriptor.getInstance().getConfig();
  private static int maxNumberOfPointsInPage;
  private static int pageSizeInByte;
  private static int groupSizeInByte;

  @BeforeClass
  public static void setUp() throws Exception {

    EnvironmentUtils.closeStatMonitor();

    // use small page setting
    // origin value
    maxNumberOfPointsInPage = tsFileConfig.getMaxNumberOfPointsInPage();
    pageSizeInByte = tsFileConfig.getPageSizeInByte();
    groupSizeInByte = tsFileConfig.getGroupSizeInByte();

    // new value
    tsFileConfig.setMaxNumberOfPointsInPage(1000);
    tsFileConfig.setPageSizeInByte(1024 * 150);
    tsFileConfig.setGroupSizeInByte(1024 * 1000);
    IoTDBDescriptor.getInstance().getConfig().setMemtableSizeThreshold(1024 * 1000);

    daemon = IoTDB.getInstance();
    daemon.active();
    EnvironmentUtils.envSetUp();

    insertData();
  }

  @AfterClass
  public static void tearDown() throws Exception {
    daemon.stop();
    // recovery value
    tsFileConfig.setMaxNumberOfPointsInPage(maxNumberOfPointsInPage);
    tsFileConfig.setPageSizeInByte(pageSizeInByte);
    tsFileConfig.setGroupSizeInByte(groupSizeInByte);
    IoTDBDescriptor.getInstance().getConfig().setMemtableSizeThreshold(groupSizeInByte);
    EnvironmentUtils.cleanEnv();
  }

  private static void insertData()
      throws ClassNotFoundException, SQLException {
    Class.forName(Config.JDBC_DRIVER_NAME);
    try (Connection connection = DriverManager
        .getConnection(Config.IOTDB_URL_PREFIX + "127.0.0.1:6667/", "root", "root");
        Statement statement = connection.createStatement()) {

      for (String sql : Constant.create_sql) {
        statement.execute(sql);
      }

      statement.execute("SET STORAGE GROUP TO root.fans");
      statement.execute("CREATE TIMESERIES root.fans.d0.s0 WITH DATATYPE=INT32, ENCODING=RLE");
      statement.execute("CREATE TIMESERIES root.fans.d0.s1 WITH DATATYPE=INT64, ENCODING=RLE");

      for (int time = 1; time < 1000; time++) {

        String sql = String
            .format("insert into root.fans.d0(timestamp,s0) values(%s,%s)", time, time % 70);
        statement.execute(sql);
        sql = String
            .format("insert into root.fans.d0(timestamp,s1) values(%s,%s)", time, time % 40);
        statement.execute(sql);
      }

    } catch (Exception e) {
      e.printStackTrace();
      fail(e.getMessage());
    }
  }

  // "select * from root.vehicle" : test select wild data
  @Test
  public void selectAllTest() throws ClassNotFoundException, SQLException {
    String selectSql = "select * from root";

    Class.forName(Config.JDBC_DRIVER_NAME);
    try (Connection connection = DriverManager
        .getConnection(Config.IOTDB_URL_PREFIX + "127.0.0.1:6667/", "root", "root")) {
      Statement statement1 = connection.createStatement();
      statement1.setFetchSize(10);
      boolean hasResultSet1 = statement1.execute(selectSql);
      Assert.assertTrue(hasResultSet1);
      ResultSet resultSet1 = statement1.getResultSet();
      int cnt = 0;
      while (resultSet1.next() && cnt < 9) {
        String ans =
            resultSet1.getString(Constant.TIMESTAMP_STR) + "," + resultSet1
                .getString("root.fans.d0.s0")
                + "," + resultSet1.getString("root.fans.d0.s1");
        cnt++;
        System.out.println(ans);
      }
      System.out.println("--------------");

      Statement statement2 = connection.createStatement();
      statement2.setFetchSize(10);
      boolean hasResultSet2 = statement2.execute(selectSql);
      Assert.assertTrue(hasResultSet2);
      ResultSet resultSet2 = statement2.getResultSet();
      cnt = 0;
      while (resultSet2.next() && cnt < 19) {
        String ans =
            resultSet2.getString(Constant.TIMESTAMP_STR) + "," + resultSet2
                .getString("root.fans.d0.s0")
                + "," + resultSet2.getString("root.fans.d0.s1");
        cnt++;
        System.out.println(ans);
      }
      System.out.println("--------------");

      cnt = 0;
      while (resultSet1.next() && cnt < 30) {
        String ans =
            resultSet1.getString(Constant.TIMESTAMP_STR) + "," + resultSet1
                .getString("root.fans.d0.s0")
                + "," + resultSet1.getString("root.fans.d0.s1");
        cnt++;
        System.out.println(ans);
      }

      resultSet1.close();
      resultSet2.close();
    } catch (
        Exception e) {
      e.printStackTrace();
      fail(e.getMessage());
    }

  }
}

```
---
But still, complement this test is good. For this mock test, I have a sugggestion that you add the following check after `statement.execute(testSql)`:
```
verify(handleIdentifier, times(1)).getQueryId()
```